File: 프로그래머스/2/87946. 피로도/피로도.cpp
#include <string>
#include <vector>
#include <algorithm>
#define MAX 8

using namespace std;

int answer = 0;
bool visited[MAX] = {0};

void bfs(int cnt, int k, vector<vector<int>> dungeons) {
    if (cnt > answer) answer = cnt;
    
    for (int i = 0; i< dungeons.size(); i++) {
        if (dungeons[i][0] <= k && !visited[i]) {
            visited[i] = 1;
            bfs(cnt + 1, k - dungeons[i][1], dungeons);
            visited[i] = 0;
        }
    }
}

int solution(int k, vector<vector<int>> dungeons) {
    bfs(0, k, dungeons);
    return answer;
}

File: 프로그래머스/2/87946. 피로도/피로도.py
from itertools import permutations

def solution(k, dungeons):
    answer = 0
    temp = k
    
    for per in permutations(dungeons):
        sum = 0
        k = temp
        for min_p, c_p in per:
            if k >= min_p:
                sum += 1
                k -= c_p
        answer = max(answer, sum)
    
    return answer